### ISSUE TITLE
GCC 11.1.1 with `-O3` causes a lot of  RuntimeWarning: invalid value . . .` errors

### ISSUE THREAD

**Issue by charris** [numpy/numpy#18949]
This is on Fedora 34 with gcc version 11.0.1. There are also a lot of zero divide warnings. They don't depend on the python version, nor even the NumPy version. Nor do they appear in python itself. There are some 60 of them in main.

```
Python 3.8.10 (default, May  8 2021, 22:00:23) 
[GCC 11.0.1 20210324 (Red Hat 11.0.1-0)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import warnings
>>> warnings.simplefilter('always')
>>> import numpy as np
>>> complex(4, 0)/4
(1+0j)
>>> np.complex128(complex(4, 0))/4
<stdin>:1: RuntimeWarning: invalid value encountered in cdouble_scalars
(1+0j)
>>> np.array(complex(4, 0))/4
<stdin>:1: RuntimeWarning: invalid value encountered in true_divide
(1+0j)
```
It may be library/compiler related.  None of these show up in the CI tests.

NumPy CPU features: SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3 AVX2 AVX512F? AVX512CD? AVX512_KNL? AVX512_KNM? AVX512_SKX? AVX512_CLX? AVX512_CNL? AVX512_ICL?

**Comment by charris** [comment]
This also happens with cpu-dispatch disabled.

python3.8 setup.py build --cpu-baseline="native" --cpu-dispatch="none"  install --user

**Comment by charris** [comment]
I'll bet it is branch prediction gone wrong in the complex division loop.

**Comment by charris** [comment]
Using `-O2` instead of `-O3` fixes the problem.

**Comment by seberg** [comment]
Maybe ping @seiko2plus and @ganesh-k13 jus tin case the division loop has some obvioius problem (i.e. doing the operation or error setting with too many values)?  Although, did we touch the `true_divide` loops recently? 

Otherwise, I guess it is possibly a gcc 11 bug...

**Comment by charris** [comment]
I think it is this code for complex division when dividing by (real) complex:
```
            if (in2r_abs == 0 && in2i_abs == 0) {
                /* divide by zero should yield a complex inf or nan */
                ((@ftype@ *)op1)[0] = in1r/in2r_abs;
                ((@ftype@ *)op1)[1] = in1i/in2i_abs;
            }
            else {
                const @ftype@ rat = in2i/in2r;
                const @ftype@ scl = 1.0@c@/(in2r + in2i*rat);
                ((@ftype@ *)op1)[0] = (in1r + in1i*rat)*scl;
                ((@ftype@ *)op1)[1] = (in1i - in1r*rat)*scl;
            }
```
When `in2i == 0`,  speculative execution of the first part of the `if` sets the fpu status register. I'm going to try `NPY_UNLIKELY` to see it that fixes it.

**Comment by seberg** [comment]
I guess `NPY_UNLIKELY` there is just as well, since it rearranges the loop code to put that (very unlikely) branch to the end of the loop and makes it tighter.  If it helps that sounds good.

Does anyone know the general way to avoid branch prediction messing with floating point error flags correctly?  Are these technicall compiler bugs or CPU bugs?

**Comment by charris** [comment]
Unfortunately, `NPY_UNLIKELY` doesn't make any difference.

**Comment by mattip** [comment]
Is this related to #11036 where we added `npy_get_floatstatus_barrier` and `npy_clear_floatstatus_barrier` to prevent compilers rearranging the order of operations?

**Comment by charris** [comment]
> Is this related to #11036 where we added

I suspect so, but was hoping to avoid that, especially as it would be awkward to do that for the ufunc loop. Hopefully the compiler will get fixed, it is supposed to handle such things. I suppose it could also be hardware related, I have old hardware.

**Comment by tacaswell** [comment]
I'm also seeing this with 

```
 22:52:28 $ gcc --version
gcc (GCC) 11.1.0
```

@charris Can you point me at how to compile with -O2?

**Comment by charris** [comment]
I use `CFLAGS`. The relevant .bashrc entry I currently use is
```
CFLAGS=" -Wall -O2 -pipe -fomit-frame-pointer"
CFLAGS=$CFLAGS" -fno-strict-aliasing -Wmaybe-uninitialized"
CFLAGS=$CFLAGS" -Wdeprecated-declarations"
CFLAGS=$CFLAGS" -Wimplicit-function-declaration"
CFLAGS=$CFLAGS" -march=native"
```
I forget why I have `-fomit-frame-pointer`.

**Comment by tacaswell** [comment]
Thank you kindly @charris !  I can also confirm that stepping back to -02 (and the rest of the flags Chuck used) eliminates the warnings for me.

**Comment by r-devulap** [comment]
Here is the disassembly of that code: https://godbolt.org/z/xW5j3nvon 

`gcc -O3 `decides to precompute the following branch 
```
if (in2r_abs >= in2i_abs) {
            if (in2r_abs == 0 && in2i_abs == 0) {
                ((double *)op1)[0] = in1r/in2r_abs;
                ((double *)op1)[1] = in1i/in2i_abs;
            }
```

by adding a division instruction `divpd %xmm2, %xmm0` (line 64 of the disassembly on the right)  prior to any of the branches and that's where the invalid is coming from.  It simply discards the result in case of branch not taken. Surely a gcc 11.1 bug and you can probably use the code snippet when you file a gcc bug. 

Note that this goes way when you use `gcc - trunk`, so may be they fixed it?

**Comment by charris** [comment]
This problem is fixed in gcc 11.2.1 (fedora 34 update).
